### PR TITLE
Copy block content in block cursor cache

### PR DESCRIPTION
**Pull request by knizhnik**
## Describe your changes

Alternative to #3740

## Issue ticket number and link

#3712

## Checklist before requesting a review
- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

### CODE REVIEW

**Comment by knizhnik** [comment]
```
main:
test_compaction.compaction: 14.814 s
test_compaction.image_creation: 6.381 s

#3740:
test_compaction.compaction: 16.845 s
test_compaction.image_creation: 6.681 s

#3748:
test_compaction.compaction: 15.440 s
test_compaction.image_creation: 6.344 s
```

File: pageserver/src/tenant/block_io.rs [hunk 1]
```diff
@@ -84,22 +84,20 @@ where
     pub fn new(reader: R) -> Self {
         BlockCursor {
             reader,
-            cache: None,
+            cache: (u32::MAX, [0u8; PAGE_SZ]),
```
Comment (koivunej) [inline pageserver/src/tenant/block_io.rs:87]: Option should be used here, then one wouldn't have to think if u32::MAX is within block numbers or not.

How do I learn more about what this block number can be...? Feels odd how could we decide within a reader that u32::MAX is in different domain than rest of the values. Perhaps we should newtype the blocknum itself to be `0..=u32::MAX-1`?

**Review by koivunej — CHANGES_REQUESTED** [review]

File: pageserver/src/tenant/block_io.rs (hunk 1 above)
Comment (knizhnik) [reply pageserver/src/tenant/block_io.rs:87]: Block with number u32::MAX corresponds to 32Tb cache size.
Not sure that we are going to support such cache in near future:)
Th emain my motivation was to eliminate one extra check (seems to be unnoticeable) and also I have some doubts about efficient copying of data to the target array when Option is used (may be also minor)

**Review by problame — CHANGES_REQUESTED** [review]
If we go with this approach, IMO we should still use (anonymous) pages from the page cache for this. 

Otherwise BlockCursors aren't accounted for in memory size limit, and we can theoretically have a lot of those in the pageserver.

**Comment by knizhnik** [comment]
#3740 is merged

**Pull request closed without merging**